core/src/test/java/com/linecorp/armeria/server/InitiateConnectionShutdownTest.java

```diff
@@ -0,0 +1,309 @@
+        inOrder.verify(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), eq(Integer.MAX_VALUE),
+                                                    eq(Http2Error.NO_ERROR.code()), eq(DEBUG_DATA));
+        inOrder.verify(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), eq(STREAM_ID),
+                                                    eq(Http2Error.NO_ERROR.code()), eq(DEBUG_DATA));
+        inOrder.verify(clientListener).onDataRead(any(ChannelHandlerContext.class), eq(STREAM_ID),
+                                                  any(ByteBuf.class), anyInt(), eq(true));
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "/goaway_async?duration=200",
+            "/goaway_blocking?duration=200",
+    })
+    void initiateConnectionShutdownCloseBeforeDrainEndHttp2(String path) throws Exception {
+        final AtomicBoolean finished = new AtomicBoolean();
+        when(clientListener.onDataRead(any(), anyInt(), any(), anyInt(), anyBoolean())).thenAnswer(
+                invocation -> {
+                    finished.set(true);
+                    return 0;
+                });
+        doAnswer((Answer<Void>) invocation -> {
+            // Retain buffer for comparison in tests.
+            final ByteBuf buf = invocation.getArgument(3);
+            buf.retain();
+            return null;
+        }).when(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), anyInt(), anyLong(),
+                                             any(ByteBuf.class));
+        clientChannel.eventLoop().execute(() -> {
+            final ChannelHandlerContext ctx = clientChannel.pipeline().firstContext();
+            final Http2Headers headers = getHttp2Headers(path);
+            http2Client.encoder().writeHeaders(ctx, STREAM_ID, headers, PADDING, false, ctx.newPromise());
+            http2Client.flush(ctx);
+        });
+        await().timeout(Duration.ofSeconds(2)).untilTrue(finished);
+        final InOrder inOrder = inOrder(clientListener);
+        inOrder.verify(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), eq(Integer.MAX_VALUE),
+                                                    eq(Http2Error.NO_ERROR.code()), eq(DEBUG_DATA));
+        inOrder.verify(clientListener).onDataRead(any(ChannelHandlerContext.class), eq(STREAM_ID),
+                                                  any(ByteBuf.class), anyInt(), eq(true));
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "/goaway_async",
+            "/goaway_async?duration=-1",
+            "/goaway_async?duration=0",
+            "/goaway_async?duration=1",
+            "/goaway_async?duration=100",
+            "/goaway_blocking",
+            "/goaway_blocking?duration=-1",
+            "/goaway_blocking?duration=0",
+            "/goaway_async?duration=1",
+            "/goaway_async?duration=100",
+    })
+    void initiateConnectionShutdownHttp1(String path) throws Exception {
```

Comment:
How about making sure all test methods in this class check if the connection is actually closed by the server, even if the client doesn't?

Reply:
Done, PTAL. Found that Armeria doesn't seem to propagate config.idleTimeoutMillis during the HTTP/2 upgrade, so it's always seem to be set to [Netty default](https://github.com/netty/netty/blob/d34212439068091bcec29a8fad4df82f0a82c638/codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2ConnectionHandlerBuilder.java#L84). Added https://github.com/line/armeria/pull/3715/files#diff-e055c2b1d925141bd5d258b22992c79d706c678dd614399284e4b5e54ea22ac3R42, please let me know if that looks reasonable to you.

Reply:
Note: discussed offline with Trustin, for HTTP/2 server connection handler we can completely disable internal Netty gracefulShutdownTimeoutMillis because drain implemented in this PR is a better way to handle graceful shutdown. This also decouples keep-alive configuration (idleTimeoutMillis) from the graceful shutdown configuration (this PR).